enterprise_gateway/proposal.md

```diff
@@ -0,0 +1,52 @@
+* Christian Kadner, IBM ('@ckadner')
+* Kun Liu, IBM ('@liukun1016')
+* Alan Chin, IBM ('@akchinSTC')
+* Sherry Guo, IBM ('@sherryxg')
+* Fred Reiss, IBM ('@frreiss')
+
+Who is the Steering Council Advocate for this Subproject?
+
+* Kyle Kelley ('@rgbkrk')
+* Peter Parente ('@parente')
+
+## Subproject goals, scope and functionality
+
+### Problem
+Founded in academia, the Jupyter projects provide a rich and highly popular set of applications for interacting with and iterating on large and complex applications.  It has been truly ground-breaking.  However, when we first attempted to build a Notebook service that could enable a large number of data scientists to run frequent and large workloads against a large Apache Spark cluster, we identified several requirements that were not currently available in the Jupyter open source ecosystem. We tried to use the Jupyter Kernel Gateway project, but we quickly realized that the JKG server became the bottleneck because the co-located Spark driver application for these kinds of workloads (in this case, the kernel process running on behalf of notebook cells) was extremely resource intensive.  Toss in a team or organization of data scientists, and you quickly saturate the compute resources of the Kernel Gateway server.  
+
+### Goals
+The primary goal behind Jupyter Enterprise Gateway is to address the gaps identified during our experience described above and provide a multi-tenant gateway that enables the ability to launch kernels as managed resources within a Spark cluster (i.e. Yarn Cluster mode).
+
+The project also aims to enhance security by providing the ability to enable encrypted communications and user impersonation of kernel processes by leveraging existing technologies such as Kerberos.
+
+By extending the kernel specifications, we plan to provide a pluggable framework for supporting other resource managers which we refer to as _process proxies_.
+
+We strive to not alter any kernel code itself.  As a result, we've developed kernel launchers that essentially house the target kernel and serve as life-cycle managers for functionality like interrupt or, in some cases, kernel termination.
+
+### Scope
+In general, Enterprise Gateway is intended to the address the needs of enterprise and cloud administrators such that cluster resource utilization is optimized across all applications, including those of data scientists performing analysis via Notebooks residing on their desktops.
+
+
+## Audience
+
+Platform Engineers and Apache Spark Administrators that want to enable Data Scientists and other Jupyter Notebook users to connect and share the Spark cluster resources when executing their notebooks.
+
+
+## Other options
+
+We are not aware of alternatives that provide the same set of capabilities, the closest that we could find is JupyterHub, but that aims to spawn multiple Jupyter Notebook servers around a given cluster - where the kernel processes are still co-located with the launching server.  
+
```

Comment:
That doesn't seem correct? JupyterHub can spin up kernels remotely via docker, docker swarm or any custom remote spawner. Am I missing something?

Reply:
Consider listing some related projects here and noting how enterprise gateway differs so that the community gets a sense of where it fits in the ecosystem. Here are a few relevant projects:

* https://github.com/danielballan/remotekernel
* https://bitbucket.org/tdaff/remote_ikernel
* https://github.com/ipython/ipyparallel
* https://github.com/korniichuk/rk

Reply:
@dhirschfeld - thank you for your comment.  
JuypyterHub spins up _Notebook servers_, while Enterprise Gateway inherits Kernel Gateway's model of spinning up _kernels_ (typically) launched on behalf of a Notebook server via [NB2KG](https://github.com/jupyter/kernel_gateway_demos/tree/master/nb2kg).  

When the kernels are remoted, they can be launched in one of two ways (although this is pluggable): 
1. As direct kernels running on hosts other than Enterprise Gateway or 
1. As Yarn cluster  _applications_ where the Yarn resource manager determines on which host the application (i.e., the kernel) is launched.  

This form of remoting places the kernel (which is typically a Spark driver application) closer to the workers, but also spread out across a cluster - thereby reducing a resource bottleneck on the kernel-launching server.  

JupyterHub could get you something similar in the Docker swarm case, but even then, the Notebook server is what is spawned, so all kernels launched from that notebook server are local and none are capable of running as actual resource-managed applications. 

@parente - great idea.  I figured there'd be JupyterHub-related questions but its good to know about these others. At quick glance, these appear to tackle the remoting at the kernelspec level whereas Enterprise Gateway provides a mechanism to plug-in _process_ lifecycle management, among other things, within the gateway itself.

Reply:
@kevin-bates I think editing the proposal to include this comment about hub vs this proposal will be useful. I too had the same initial reaction as @dhirschfeld

Reply:
I'm just a little concerned about fracturing the ecosystem and having two projects which do very similar things meaning development of either would be slower than if everyone pooled their resources. Everyone is free to scratch their own itch but I'm still unconvinced that JupyterHub couldn't easily satisfy the requirements above with much less effort than a whole new project.

@yuvipanda or @minrk would probably know better that me but it seems that it would be pretty easy to get JupyterHub spinning up kernels rather than notebook servers. In my setup JupyterHub spins up a remote docker container with Anaconda installed and my config tells JupyterHub to pass `jupyter labhub *args` to the container. To spin up a remote kernel instead I would just change the container command to `ipykernel *args`. That *might* require some minor changes to JupyterHub to not interact with the kernel, I'm not sure, but I think it's certainly possible. As for yarn, it seems that could potentially be supported with a custom `YarnSpawner`

With the new named servers feature in JupyterHub I can imagine it being capable of managing remote kernels for not just notebook instances in the browser but desktop apps such as JupyterLab, nteract or even Spyder. @minrk might have a thing or two to say about scope creep but that was the direction I thought things were heading...

Anyway, just wanted to make sure all options had been properly considered and dismissing JupyterHub as incapable of supporting remote kernels seemed to be starting from a questionable premise.

Reply:
@dirschfeld, Thanks for taking the time to review and comment on the proposal. As @rgbkrk mentioned in his first comment, I see the “Enterprise Gateway” as really being an enhancement to the existing Kernel Gateway, and the target audience being Bring-Your-Own-Notebook where data scientists run notebooks on their desktops - on the other side of a firewall - and thus need a “gateway” to enable sharing the Spark runtime, which to me is a very specific scenario, which to certain extent is common in enterprises or cloud environments. In this case, if JKG wasn’t fracturing the JupyterHub community, I am certain that Enterprise Gateway will not either.

Reply:
Sure, if it's useful and fills a niche that's great. I'm not sure about the name "Enterprise" though - to me Spark *isn't* synonymous with Enterprise so perhaps a more targeted name would be appropriate?

For background, I'm deploying JupyterHub in the "Enterprise", using it to spin up JupyterLab running in containers on our Windows HPC cluster. My hope is that I will be able to also use the same infrastructure to support spinning up remote kernels for any application but JupyterLab, nteract and spyder in particular as those are the ones we're using day-to-day.

Reply:
@dhirschfeld The name chosen was "Enterprise Gateway" to distinguish the project from "Kernel Gateway". It's mostly an adjective to the gateway noun and the "enterprise" version provides extra functionality to a gateway that is usually seen on enterprise/cloud deployments.